docs/architecture/adr-001-coin-source-tracing.md

```diff
@@ -239,117 +271,131 @@ func GetTraceHashFromDenom(rawDenom string) (tmbytes.HexBytes, error) {
-
 ```golang
-// createOutgoingPacket
+// SendTransfer
 // ...
-prefix := types.GetDenomPrefix(destinationPort, destinationChannel)
-source := strings.HasPrefix(amount[0].Denom, prefix)
 
-for _, denomTrace := range denomTraces {
+// NOTE: SendTransfer simply sends the denomination as it exists on its own
+// chain inside the packet data. The receiving chain will perform denom
+// prefixing as necessary.
+
+if types.SenderChainIsSource(sourcePort, sourceChannel, token.Denom) {
+  // create the escrow address for the tokens
+  escrowAddress := types.GetEscrowAddress(sourcePort, sourceChannel)
+
+  // escrow source tokens. It fails if balance insufficient.
+  if err := k.bankKeeper.SendCoins(
+    ctx, sender, escrowAddress, sdk.NewCoins(token),
+  ); err != nil {
+    return err
+  }
+} else {
   // set the value to the lookup table if not stored already
   traceHash := denomTrace.Hash()
   if !k.HasDenomTrace(ctx, traceHash) {
     k.SetDenomTrace(ctx, traceHash, denomTrace)
   }
-}
 
-if source {
-  // clear the denomination from the prefix to send the coins to the escrow account
-  coins := make(sdk.Coins, len(amount))
-  for i, coin := range amount {
-    if strings.HasPrefix(denomTraces[i].Trace, prefix) {
-      if err := denomTraces[i].RemovePrefix(); err != nil {
-        return err
-      }
-
-      // set the new value to the lookup table if not stored already
-      traceHash := denomTraces[i].Hash()
-      if !k.HasDenomTrace(ctx, traceHash) {
-        k.SetDenomTrace(ctx, traceHash, denomTrace[i])
-      }
-
-      coins[i] = sdk.NewCoin(denomTrace[i].IBCDenom(), coin.Amount)
-    } else {
-      coins[i] = coin
-    }
+  // transfer the coins to the module account and burn them
+  if err := k.bankKeeper.SendCoinsFromAccountToModule(
+    ctx, sender, types.ModuleName, sdk.NewCoins(token),
+  ); err != nil {
+    return err
   }
 
-  // escrow tokens if the destination chain is the same as the sender's
-  escrowAddress := types.GetEscrowAddress(sourcePort, sourceChannel)
-
-  // escrow source tokens. It fails if balance insufficient.
-  if err := k.bankKeeper.SendCoins(
-    ctx, sender, escrowAddress, coins,
+  if err := k.bankKeeper.BurnCoins(
+    ctx, types.ModuleName, sdk.NewCoins(token),
   ); err != nil {
+    // NOTE: should not happen as the module account was
+    // retrieved on the step above and it has enough balance
+    // to burn.
     return err
   }
-} else {
-  // build the receiving denomination prefix if it's not present
-  prefix = types.GetDenomPrefix(sourcePort, sourceChannel)
-  for _, coin := range amount {
-    if !strings.HasPrefix(denomTrace[i].Trace, prefix) {
-      return Wrapf(types.ErrInvalidDenomForTransfer, "invalid token denom trace prefix: %s", denomTrace[i].IBCDenom())
-    }
-  }
-// ...
 }
+// ...
 ```
 
+The denomination trace info also needs to be updated when token is received in both cases:
+
+- Sender is **sink** chain: Store the received denomination, i.e in the [example](#example) above,
```

Comment:
We don't need to update the denomination trace info on receiving chain is receiver is the source since the trace should already exist in store

Reply:
>  is receiver is the source since the trace should already exist in store

I didn't understand this part

Reply:
*if the receiver is the source since the trace should already exist in store

@AdityaSripal is saying if the receiver is the source, then the receiver created the token and must have the trace lookup already stored (if necessary ie native token case wouldn't need a lookup)

Reply:
I think the what is currently written is conflicting logic:

`The denomination trace info only needs to be updated when token is received, in both scenarios:`

conflicts with

`As there the trimmed trace info is already known by the chain we don't need to store it (i.e 'C/B/A/denom').`